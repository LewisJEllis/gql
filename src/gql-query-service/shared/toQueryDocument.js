/* @flow */
import { Source } from 'graphql/language/source';
import whileSafe from 'gql-shared/whileSafe';
import debug from 'gql-shared/debug';
import { type IParser } from 'gql-shared/types';
import MultilineCharacterStream from 'gql-shared/MultilineCharacterStream';
import { EmbeddedLanguageParser } from 'gql-shared/EmbeddedLanguageParser';
import invariant from 'invariant';

const IRREGULAR_WHITESPACE =
  '\f\v\u0085\u00A0\ufeff\u00a0\u1680\u180e\u2000\u2001\u2002\u2003\u2004\u2005\u2006\u2007\u2008\u2009\u200a\u200b\u202f\u205f\u3000';

const whiteSpaceString = text =>
  text.replace(new RegExp(`[\\S${IRREGULAR_WHITESPACE}]`, 'g'), ' ');

function placeholderFragment(text: string) {
  const fragmentName = 'F'; // dummy frag name
  const str = text.replace(/[^\s]/g, ',');
  const fragmentStr = `...${fragmentName}`;
  if (/^.{2}\n/.test(str)) {
    // ${
    //    fragment
    //  }
    return str.replace(/^.{2}/, fragmentStr);
  }
  return str.replace(new RegExp(`^.{${fragmentStr.length}}`), fragmentStr);
}

export default function toQueryDocument(
  source: Source,
  parser: IParser,
): Array<string> {
  debug.time('toQueryDocument');
  const state = parser.startState();

  const stream = new MultilineCharacterStream(source.body);
  const isEmbeddedMode = parser instanceof EmbeddedLanguageParser;
  const queryDocuments = [];
  let activeQueryDocument = isEmbeddedMode ? null : ''; // using null in embedded to differentiate b/w no tag vs empty tag query

  whileSafe(
    {
      condition: () => stream.getCurrentPosition() < source.body.length,
      call: () => {
        const style = parser.token(stream, state);
        // console.log('current', `[${stream.current()}]`, style);

        if (isEmbeddedMode && style === 'ws-empty') {
          return;
        }

        if (isEmbeddedMode && style === 'ws-start') {
          // NOTE: each document should retain the position
          // so should prepend with all previous string
          const emptyPreviousDocument =
            queryDocuments.length > 0
              ? whiteSpaceString(queryDocuments[queryDocuments.length - 1])
              : '';

          activeQueryDocument =
            emptyPreviousDocument + whiteSpaceString(stream.current());
          return;
        }

        if (activeQueryDocument === null) {
          // Somethings is wrong
          invariant(false, 'activeQueryDocument should not be null here');
        }

        // embedded documents
        if (isEmbeddedMode && style === 'ws-end') {
          activeQueryDocument += whiteSpaceString(stream.current());
          queryDocuments.push(activeQueryDocument);
          // reset
          activeQueryDocument = null;
          return;
        }

        if (style === 'ws-2') {
          activeQueryDocument += whiteSpaceString(stream.current());
          return;
        }

        if (
          // add fragment name is missing
          parser.options.allowFragmentWithoutName &&
          state.kind === 'TypeCondition' &&
          state.prevState.kind === 'FragmentDefinition' &&
          stream.current() === 'on' &&
          !state.prevState.name
        ) {
          activeQueryDocument += '_ on';
          return;
        }

        if (style === 'js-frag') {
          activeQueryDocument += placeholderFragment(stream.current());
          return;
        }

        if (style) {
          activeQueryDocument += stream.current();
        }
      },
    },
    source.body.length,
  );

  if (activeQueryDocument !== null) {
    queryDocuments.push(activeQueryDocument);
  }

  debug.timeEnd('toQueryDocument');
  // console.log(queryDocuments);
  return queryDocuments;
}
