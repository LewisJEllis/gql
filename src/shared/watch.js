/* @flow */
import { type WatchFile } from './types';
import { type FileMatchConfig } from '../config/GQLConfig';
import sane from 'sane';
import globby from 'globby';
import _debounce from 'lodash/debounce';
import parseFileMatchConfig from './parseFileMatchConfig';

class WatchEventsBatcher {
  _listener: (files: Array<WatchFile>) => void;
  _queue = [];
  _activeTimeout = null;
  _waitTimeForBatching = 200 /* msec */;

  constructor(listener: (files: Array<WatchFile>) => void) {
    this._listener = listener;
  }

  __dispatchQueue = _debounce(() => {
    const events = this._queue;
    this._queue = [];

    if (events.length === 0) {
      return;
    }

    this._listener(events);
  }, this._waitTimeForBatching);

  add(eventOrEvents: WatchFile | Array<WatchFile>) {
    this._queue.push(...(Array.isArray(eventOrEvents) ? eventOrEvents : [eventOrEvents]));
    this.__dispatchQueue();
  }
}

export type WatchOptions = {
  rootPath: string,
  files: FileMatchConfig,
  name: string,
  watchman?: boolean,
  watch?: boolean,
  onChange: (changedFiles: Array<WatchFile>) => void,
};

function boolValue(value: ?boolean, defaultValue: boolean) {
  return typeof value === 'boolean' ? value : defaultValue;
}

function setupWatcher({ rootPath, onChange, name, glob, ignored, watchman }) {
  const watchEventsBatcher = new WatchEventsBatcher(onChange);
  const useWatchman = boolValue(watchman, true);
  const watcherName = useWatchman ? 'watchman-watcher' : 'node-watcher';

  const watcher = sane(rootPath, {
    glob,
    watchman: useWatchman,
    ignored,
  });

  console.log(`[${watcherName}] '${name}' starting...`);
  const key = `[${watcherName}] '${name}' started`;
  console.time(key);
  watcher.on('ready', () => {
    console.timeEnd(key);
  });

  watcher.on('change', (filePath) => {
    watchEventsBatcher.add({ name: filePath, exists: true });
  });

  watcher.on('add', (filePath) => {
    watchEventsBatcher.add({ name: filePath, exists: true });
  });

  watcher.on('delete', (filePath) => {
    watchEventsBatcher.add({ name: filePath, exists: false });
  });

  return watcher;
}

export default function watch(options: WatchOptions) {
  const { glob, ignored } = parseFileMatchConfig(options.files);

  let watcher = null;
  globby(glob, { cwd: options.rootPath })
    .then((matches) => {
      const watchFiles = matches.map((file) => ({ name: file, exists: true }));
      if (!boolValue(options.watch, true)) {
        return watchFiles;
      }
      // if in watchmode then wait for watcher to start
      return new Promise((resolve) => {
        watcher = setupWatcher({
          rootPath: options.rootPath,
          onChange: options.onChange,
          watchman: options.watchman,
          name: options.name,
          glob,
          ignored,
        });
        watcher.on('ready', resolve);
      }).then(() => watchFiles);
    })
    .then((watchFiles) => {
      options.onChange(watchFiles);
    })
    .catch((err) => {
      throw err;
    });

  return {
    close() {
      if (watcher) {
        watcher.close();
      }
    },
  };
}
