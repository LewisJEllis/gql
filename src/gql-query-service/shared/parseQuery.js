/* @flow */
import { Source } from 'graphql/language/source';
import { type DocumentNode } from 'graphql/language/ast';
import { parse as parseGraphqlDocument } from 'graphql/language/parser';
import { type IParser } from 'gql-shared/types';
import { toGQLError, type GQLError, SEVERITY } from 'gql-shared/GQLError';
import toQueryDocument from './toQueryDocument';

export type ParsedQuery = {|
  queries: Array<{|
    ast: ?DocumentNode,
    error: ?GQLError,
    sourceText: string,
  |}>,
  isEmpty: boolean,
|};

export default function parserQuery(
  source: Source,
  parser: IParser,
): ParsedQuery {
  const queryDocuments = toQueryDocument(source, parser);
  return {
    isEmpty: queryDocuments.length === 0,
    queries: queryDocuments.map(doc => {
      try {
        const ast = parseGraphqlDocument(new Source(doc, source.name));
        return {
          ast,
          error: null,
          sourceText: doc,
        };
      } catch (err) {
        return {
          ast: null,
          error: toGQLError(err, SEVERITY.error),
          sourceText: doc,
        };
      }
    }),
  };
}
