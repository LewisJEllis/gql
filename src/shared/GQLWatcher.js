/* @flow */
import watch, { type WatchOptions } from './watch';
import { execSync } from 'child_process';
import { dedent } from 'dentist';

type Options = {
  watchman?: boolean,
  watch?: boolean,
};

export default class GQLWatcher {
  _watchers: Array<*> = [];
  _options: *;

  constructor(options: Options) {
    this._options = {
      ...options,
      watchman: typeof options.watchman === 'boolean' ? options.watchman : true,
      watch: typeof options.watch === 'boolean' ? options.watch : true,
    };

    if (this._options.watchman && this._options.watch) {
      this._verifyWatchmanInstalled();
    }
  }

  watch(options: WatchOptions) {
    const watcher = watch({
      ...options,
      watchman: this._options.watchman,
      watch: this._options.watch,
    });
    this._watchers.push(watcher);
  }

  close() {
    this._watchers.forEach((watcher) => {
      watcher.close();
    });
  }

  _verifyWatchmanInstalled = () => {
    try {
      execSync('watchman --version', { stdio: ['ignore'] });
    } catch (e) {
      const error = new Error(
        dedent(`
          Watchman was not found in PATH. See
          'https://facebook.github.io/watchman/docs/install.html
          for installation instructions.

          If you dont want to use watchman set
          "watchman: false" in options
        `),
      );
      error.stack = ''; // dont want stack track to show up
      throw error;
    }
  };
}
