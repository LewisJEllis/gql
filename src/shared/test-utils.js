/* @flow */
import os from 'os';
import objectHash from 'object-hash';
import { writeFileSync } from '../shared/fs-utils';
import path from 'path';

export const TEST_TEMP_DIR = path.join(os.tmpdir(), 'gql-test');

export function createTempFiles(files: { [file: string]: string }): string {
  // deterministic dirName [as this path used in test cases output (see generated snapshot)]
  const dir = path.join(TEST_TEMP_DIR, objectHash(files));
  Object.keys(files).forEach((file) => {
    const content = files[file];
    writeFileSync(path.join(dir, file), content);
  });
  return dir;
}
