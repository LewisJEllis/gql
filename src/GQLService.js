/* @flow */
import * as schema from './schema';
import * as query from './query';

import GQLConfig from './config/GQLConfig';
import GQLWatcher from './shared/GQLWatcher';
import debug from './shared/debug';

import type { GQLHint, GQLInfo, DefLocation, Position } from './shared/types';

import type { GQLError } from './shared/GQLError';

type Options = $Exact<{
  cwd?: string,
  onChange?: () => void,
  onInit?: () => void,
  debug?: boolean,
  watchman?: boolean,
  watch?: boolean,
}>;

type CommandParams = $Exact<{
  sourcePath: string,
  sourceText: string,
  position: Position,
}>;

export type { Options as GQLServiceOptions };

const _noop = () => {}; // eslint-disable-line no-empty-function

export class GQLService {
  _isInitialized: boolean = false;
  _schemaManager: schema.SchemaManager;
  _queryManager: ?query.QueryManager;

  _config: GQLConfig;

  constructor(options: ?Options) {
    const {
      onInit = _noop,
      onChange = _noop,
      watchman,
      watch,
      debug: enableDebug,
      ...configOptions
    } = options || {};
    if (enableDebug) {
      (debug: any).enable();
    }

    this._config = new GQLConfig(configOptions);
    const watcher = new GQLWatcher({ watchman, watch });

    this._schemaManager = new schema.SchemaManager({
      config: this._config,
      watcher,
    });

    this._schemaManager.onChange(onChange);
    this._schemaManager.onInit(() => {
      if (!this._config.getQuery()) {
        this._isInitialized = true;
        onInit();
        return;
      }

      const queryManager = new query.QueryManager({
        config: this._config,
        getSchema: () => this._schemaManager.getSchema(),
        watcher,
      });

      this._queryManager = queryManager;

      queryManager.onInit(() => {
        this._isInitialized = true;
        onInit();
      });
      queryManager.onChange(onChange);
    });
  }

  // list all file extensions defined in gqlconfig
  getFileExtensions(): Array<string> {
    return this._config.getFileExtensions();
  }

  getSchema() {
    return this._schemaManager.getSchema();
  }

  status(): Array<GQLError> {
    if (!this._isInitialized) {
      return [];
    }
    const schemaErrors = this._schemaManager.getSchemaErrors();
    const queryErrors = this._queryManager ? this._queryManager.getErrors() : [];
    return schemaErrors.concat(queryErrors.filter((err) => Boolean(err)));
  }

  autocomplete(params: CommandParams): Array<GQLHint> {
    debug.log('\n[autocomplete]');
    debug.time('time');
    const { sourceText, sourcePath, position } = params;
    if (!this._isInitialized) {
      return [];
    }

    // codemirror instance
    let results = [];
    debug.time('match');
    const match = this._config.match(sourcePath);
    debug.timeEnd('match');
    debug.log('FileType:', match && match.type);

    debug.time('autocomplete');
    if (match && match.type === 'schema') {
      results = schema.commands.getHintsAtPosition(
        this._schemaManager.getSchema(),
        sourceText,
        position,
      );
    }

    if (match && match.type === 'query') {
      results = query.commands.getHintsAtPosition(
        this._schemaManager.getSchema(),
        sourceText,
        position,
        match.opts,
      );
    }
    debug.timeEnd('autocomplete');
    debug.timeEnd('time');
    return results;
  }

  getDef(params: CommandParams): ?DefLocation {
    debug.log('\n[getDef]');
    debug.time('time');
    const { sourceText, sourcePath, position } = params;
    if (!this._isInitialized) {
      return undefined;
    }

    let defLocation = null;
    debug.time('match');
    const match = this._config.match(sourcePath);
    debug.timeEnd('match');
    debug.log('FileType:', match && match.type);

    debug.time('getDef');
    if (match && match.type === 'schema') {
      defLocation = schema.commands.getDefinitionAtPosition(
        this._schemaManager.getSchema(),
        sourceText,
        position,
      );
    }

    if (match && match.type === 'query') {
      defLocation = query.commands.getDefinitionAtPosition(
        this._schemaManager.getSchema(),
        sourceText,
        position,
        match.opts.parser,
      );
    }
    debug.timeEnd('getDef');
    debug.timeEnd('time');
    return defLocation;
  }

  findRefs(params: CommandParams): Array<DefLocation> {
    debug.log('\n[findRefs]');
    debug.time('time');
    const { sourceText, sourcePath, position } = params;
    if (!this._isInitialized) {
      return [];
    }

    let refLocations = [];
    debug.time('match');
    const match = this._config.match(sourcePath);
    debug.timeEnd('match');
    debug.log('FileType:', match && match.type);

    debug.time('findRefs');
    if (match && match.type === 'schema') {
      refLocations = schema.commands.findRefsOfTokenAtPosition(
        this._schemaManager.getSchema(),
        sourceText,
        position,
      );
    }
    debug.timeEnd('findRefs');

    debug.timeEnd('time');
    // @TODO query not implemented
    return refLocations;
  }

  getInfo(params: CommandParams): ?GQLInfo {
    debug.log('\n[getInfo]');
    debug.time('time');
    const { sourcePath, sourceText, position } = params;
    if (!this._isInitialized) {
      return undefined;
    }

    let info = null;
    debug.time('match');
    const match = this._config.match(sourcePath);
    debug.timeEnd('match');
    debug.log('FileType:', match && match.type);

    debug.time('getInfoAtToken');
    if (match && match.type === 'schema') {
      info = schema.commands.getInfoOfTokenAtPosition(
        this._schemaManager.getSchema(),
        sourceText,
        position,
      );
    }

    if (match && match.type === 'query') {
      info = query.commands.getInfoOfTokenAtPosition(
        this._schemaManager.getSchema(),
        sourceText,
        position,
        match.opts,
      );
    }
    debug.timeEnd('getInfoAtToken');

    debug.timeEnd('time');
    return info;
  }
}
