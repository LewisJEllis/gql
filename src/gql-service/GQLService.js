/* @flow */
import GQLSchemaService from 'gql-schema-service';
import GQLQueryService from 'gql-query-service';
import GQLConfig from 'gql-config';
import GQLWatcher from 'gql-watcher';
import GQLBaseService from 'gql-shared/GQLBaseService';

import debug from 'gql-shared/debug';

import {
  type GQLHint,
  type GQLInfo,
  type GQLLocation,
  type GQLPosition,
} from 'gql-shared/types';

import { type GQLError } from 'gql-shared/GQLError';

type Options = $Exact<{
  cwd?: string,

  // watch
  watchman?: boolean,
  watch?: boolean,

  // debug
  debug?: boolean,
}>;

type CommandParams = $Exact<{
  sourcePath: string,
  sourceText: string,
  position: GQLPosition,
}>;

export type { Options as GQLServiceOptions };

export default class GQLService extends GQLBaseService {
  _schemaService: GQLSchemaService;
  _queryService: ?GQLQueryService;

  _config: GQLConfig;
  _watcher: GQLWatcher;

  constructor(_options: ?Options) {
    super();
    const options = _options || {};

    // enable debug mode
    if (options.debug) {
      debug.enable();
    }

    this._config = new GQLConfig({ cwd: options.cwd || process.cwd() });
    this._watcher = new GQLWatcher({
      watchman: options.watchman,
      watch: options.watch,
    });

    // setup schema service
    this._schemaService = new GQLSchemaService({
      config: this._config,
      watcher: this._watcher,
    });
    this._schemaService.onChange(this._triggerChange);
    this._schemaService.onError(this._triggerError);

    // setup query service
    if (this._config.getQueryConfig()) {
      const queryService = new GQLQueryService({
        config: this._config,
        getSchema: () => this._schemaService.getSchema(),
        watcher: this._watcher,
      });
      queryService.onChange(this._triggerChange);
      queryService.onError(this._triggerError);
      this._queryService = queryService;
    }
  }

  async _handleStart() {
    // start schema service
    await this._schemaService.start();
    if (this._queryService) {
      await this._queryService.start();
    }
  }

  async _handleStop() {
    await this._schemaService.stop();
    if (this._queryService) {
      await this._queryService.stop();
    }
  }

  getSchema(): GQLSchema {
    return this._schemaService.getSchema();
  }

  getConfig(): GQLConfig {
    return this._config;
  }

  status(): Array<GQLError> {
    if (!this._isRunning) {
      return [];
    }
    try {
      const schemaErrors = this._schemaService.getSchemaErrors();
      const queryErrors = this._queryService
        ? this._queryService.getErrors()
        : [];
      return schemaErrors.concat(queryErrors.filter(err => Boolean(err)));
    } catch (err) {
      this._triggerError(err);
      return [];
    }
  }

  autocomplete(params: CommandParams): Array<GQLHint> {
    return this._catchThrownErrors(() => {
      debug.log('\n[autocomplete]');
      debug.time('time');
      if (!this._isRunning) {
        return [];
      }

      // codemirror instance
      let results = [];
      debug.time('fileConfig');
      const fileConfig = this._config.getFileConfig(params.sourcePath);
      debug.timeEnd('fileConfig');
      debug.log('FileType:', fileConfig && fileConfig.type);

      debug.time('autocomplete');
      if (fileConfig && fileConfig.type === 'schema') {
        results = this._schemaService.getHintsAtPosition({
          fileContent: params.sourceText,
          fileOptions: fileConfig.opts,
          position: params.position,
        });
      }
      if (this._queryService && fileConfig && fileConfig.type === 'query') {
        results = this._queryService.getHintsAtPosition({
          fileContent: params.sourceText,
          fileOptions: fileConfig.opts,
          position: params.position,
        });
      }
      debug.timeEnd('autocomplete');
      debug.timeEnd('time');
      return results;
    }, []);
  }

  getDef(params: CommandParams): ?GQLLocation {
    return this._catchThrownErrors(() => {
      debug.log('\n[getDef]');
      debug.time('time');
      if (!this._isRunning) {
        return null;
      }

      let defLocation = null;
      debug.time('fileConfig');
      const fileConfig = this._config.getFileConfig(params.sourcePath);
      debug.timeEnd('fileConfig');
      debug.log('FileType:', fileConfig && fileConfig.type);

      debug.time('getDef');
      if (fileConfig && fileConfig.type === 'schema') {
        defLocation = this._schemaService.getDefinitionAtPosition({
          fileContent: params.sourceText,
          fileOptions: fileConfig.opts,
          position: params.position,
        });
      }

      if (this._queryService && fileConfig && fileConfig.type === 'query') {
        defLocation = this._queryService.getDefinitionAtPosition({
          fileContent: params.sourceText,
          fileOptions: fileConfig.opts,
          position: params.position,
        });
      }
      debug.timeEnd('getDef');
      debug.timeEnd('time');
      return defLocation;
    }, null);
  }

  findRefs(params: CommandParams): Array<GQLLocation> {
    return this._catchThrownErrors(() => {
      debug.log('\n[findRefs]');
      debug.time('time');
      if (!this._isRunning) {
        return [];
      }

      let refLocations = [];
      debug.time('fileConfig');
      const fileConfig = this._config.getFileConfig(params.sourcePath);
      debug.timeEnd('fileConfig');
      debug.log('FileType:', fileConfig && fileConfig.type);

      debug.time('findRefs');
      if (fileConfig && fileConfig.type === 'schema') {
        refLocations = this._schemaService.findRefsOfTokenAtPosition({
          fileContent: params.sourceText,
          fileOptions: fileConfig.opts,
          position: params.position,
        });
      }
      debug.timeEnd('findRefs');

      debug.timeEnd('time');
      // @TODO query not implemented
      return refLocations;
    }, []);
  }

  getInfo(params: CommandParams): ?GQLInfo {
    return this._catchThrownErrors(() => {
      debug.log('\n[getInfo]');
      debug.time('time');
      if (!this._isRunning) {
        return null;
      }

      let info = null;
      debug.time('fileConfig');
      const fileConfig = this._config.getFileConfig(params.sourcePath);
      debug.timeEnd('fileConfig');
      debug.log('FileType:', fileConfig && fileConfig.type);

      debug.time('getInfoAtToken');
      if (fileConfig && fileConfig.type === 'schema') {
        info = this._schemaService.getInfoOfTokenAtPosition({
          fileContent: params.sourceText,
          fileOptions: fileConfig.opts,
          position: params.position,
        });
      }

      if (this._queryService && fileConfig && fileConfig.type === 'query') {
        info = this._queryService.getInfoOfTokenAtPosition({
          fileContent: params.sourceText,
          fileOptions: fileConfig.opts,
          position: params.position,
        });
      }
      debug.timeEnd('getInfoAtToken');

      debug.timeEnd('time');
      return info;
    }, null);
  }
}
