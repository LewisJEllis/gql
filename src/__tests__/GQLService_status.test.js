/* @flow */
/* eslint-disable max-len */
import { GQLService } from '../GQLService';
import { createTempFiles } from '../shared/test-utils';

describe('Schema', () => {
  it('should report errors in schema', (done) => {
    const gql = new GQLService({
      cwd: createTempFiles({
        '.gqlconfig': `
          {
            schema: {
              files: 'schema/*.gql',
            }
          }
        `,
        'schema/schema.gql': `
          type Query {
            viewer: xViewer # missing viewer
          }
        `,
      }),
      watch: false,
      onInit() {
        expect(gql.status()).toMatchSnapshot();
        done();
      },
    });
  });

  it('can modify validation rule severity', (done) => {
    const gql = new GQLService({
      cwd: createTempFiles({
        'schema/schema.gql': `
          type Query {
            name: String
          }
          type Hello {
            name: String
          }
        `,
        '.gqlconfig': `{
          schema: {
            files: 'schema/*.gql',
            validate: {
              extends: 'gql-rules-schema',
              rules: {
                NoUnusedTypeDefinition: 'error',
              },
            }
          }
        }`,
      }),
      watch: false,
      onInit: () => {
        expect(gql.status()).toMatchSnapshot();
        done();
      },
    });
  });

  it('calling status before onInit should not throw', (done) => {
    const gql = new GQLService({
      cwd: createTempFiles({
        '.gqlconfig': `
          {
            schema: {
              files: 'schema/*.gql',
            }
          }
        `,
        'schema/schema.gql': `
          type Query {
            viewer: xViewer # missing viewer
          }
        `,
      }),
      watch: false,
      onInit: done,
    });
    expect(() => gql.status()).not.toThrow();
    expect(gql.status()).toEqual([]);
  });

  it('can turn off validation rules', (done) => {
    const gql = new GQLService({
      cwd: createTempFiles({
        'schema/schema.gql': `
            type Query {
              name: String
            }
            type Hello {
              name: String
            }
          `,
        '.gqlconfig': `{
            schema: {
              files: 'schema/*.gql',
              validate: {
                extends: 'gql-rules-schema',
                rules: {
                  NoUnusedTypeDefinition: 'off',
                },
              }
            }
          }`,
      }),
      watch: false,
      onInit: () => {
        expect(gql.status()).toMatchSnapshot();
        done();
      },
    });
  });
});

describe('Query', () => {
  it('report query errors', (done) => {
    const gql = new GQLService({
      cwd: createTempFiles({
        '.gqlconfig': `
          {
            schema: {
              files: 'schema/*.gql',
            },
            query: {
              files: [
                {
                  match: 'query/*.gql',
                  parser: 'QueryParser',
                }
              ]
            }
          }
        `,
        'schema/schema.gql': `
          type Query {
            viewer: Viewer
          }
          type Viewer {
            name: String
          }
        `,
        'query/query.gql': `
          query {
            xViewer {
              name
            }
          }
        `,
      }),
      watch: false,
      onInit: () => {
        expect(gql.status()).toMatchSnapshot();
        done();
      },
    });
  });
});
