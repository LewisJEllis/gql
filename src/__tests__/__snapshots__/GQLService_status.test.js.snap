// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Query report query errors 1`] = `
Array [
  Object {
    "locations": Array [
      Object {
        "column": 13,
        "line": 3,
        "path": "/tmp/gql-test/c57fb271cbb161eb984617f3e9dff63ff8a3dab6/query/query.gql",
      },
    ],
    "message": "Cannot query field \\"xViewer\\" on type \\"Query\\". Did you mean \\"viewer\\"? (FieldsOnCorrectType)",
    "severity": "error",
  },
]
`;

exports[`Schema can modify validation rule severity 1`] = `
Array [
  Object {
    "locations": Array [
      Object {
        "column": 11,
        "line": 5,
        "path": "/tmp/gql-test/a1c3b7f4d95aae42520a0beaaebd08968b164ca2/schema/schema.gql",
      },
    ],
    "message": "Unused type definition 'Hello' (NoUnusedTypeDefinition)",
    "severity": "error",
  },
]
`;

exports[`Schema can turn off validation rules 1`] = `Array []`;

exports[`Schema should report errors in schema 1`] = `
Array [
  Object {
    "locations": Array [
      Object {
        "column": 21,
        "line": 3,
        "path": "/tmp/gql-test/ca7b65d03c4b6f03d11204e53a0c1a7e510c5b29/schema/schema.gql",
      },
    ],
    "message": "Type \\"xViewer\\" not found.",
    "severity": "error",
  },
]
`;
